File: contents/split-operator_method/code/c/split_op.c
#include <complex.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

// Using fftw3 library.
#include <fftw3.h>

struct params {
    double xmax;
    unsigned int res;
    double dt;
    unsigned int timesteps;
    double dx;
    double *x;
    double dk;
    double *k;
    bool im_time;
};

struct operators {
    size_t size;
    double complex *v;
    double complex *pe;
    double complex *ke;
    double complex *wfc;
};

void fft(double complex *x, int n, bool inverse) {
    double complex y[n];
    memset(y, 0, sizeof(y));
    fftw_plan p;

    if (inverse) {
        p = fftw_plan_dft_1d(n, (fftw_complex*)x, (fftw_complex*)y,
                             FFTW_BACKWARD, FFTW_ESTIMATE);
    } else {
        p = fftw_plan_dft_1d(n, (fftw_complex*)x, (fftw_complex*)y,
                             FFTW_FORWARD, FFTW_ESTIMATE);
    }

    fftw_execute(p);
    fftw_destroy_plan(p);

    for (size_t i = 0; i < n; ++i) {
        x[i] = y[i] / sqrt((double)n);
    }
}

void init_params(struct params *par, double xmax, unsigned int res, double dt,
                 unsigned int timesteps, bool im) {

    par->xmax = xmax;
    par->res = res;
    par->dt = dt;
    par->timesteps = timesteps;
    par->dx = 2.0 * xmax / res;
    par->x = malloc(sizeof(double) * res);
    par->dk = M_PI / xmax;
    par->k = malloc(sizeof(double) * res);
    par->im_time = im;

    for (size_t i = 0; i < res; ++i) {
        par->x[i] = xmax / res - xmax + i * (2.0 * xmax / res);
        if (i < res / 2) {
            par->k[i] = i * M_PI / xmax;
        } else {
            par->k[i] = ((double)i - res) * M_PI / xmax;
        }
    }
}

void init_operators(struct operators *opr, struct params par, double voffset,
                    double wfcoffset) {

    opr->size = par.res;
    opr->v = malloc(sizeof(double complex) * par.res);
    opr->pe = malloc(sizeof(double complex) * par.res);
    opr->ke = malloc(sizeof(double complex) * par.res);
    opr->wfc = malloc(sizeof(double complex) * par.res);

    for (size_t i = 0; i < par.res; ++i) {
        opr->v[i] = 0.5 * cpow(par.x[i] - voffset, 2);
        opr->wfc[i] = cexp(-cpow(par.x[i] - wfcoffset, 2) / 2.0);

        if (par.im_time) {
            opr->ke[i] = cexp(-0.5 * par.dt * cpow(par.k[i], 2));
            opr->pe[i] = cexp(-0.5 * par.dt * opr->v[i]);
        } else {
            opr->ke[i] = cexp(-0.5 * par.dt * cpow(par.k[i], 2) * I);
            opr->pe[i] = cexp(-0.5 * par.dt * opr->v[i] * I);
        }
    }
}

void split_op(struct params par, struct operators opr) {
    double density[opr.size];

    for (size_t i = 0; i < par.timesteps; ++i) {
        for (size_t j = 0; j < opr.size; ++j) {
            opr.wfc[j] *= opr.pe[j];
        }

        fft(opr.wfc, opr.size, false);

        for (size_t j = 0; j < opr.size; ++j) {
            opr.wfc[j] *= opr.ke[j];
        }

        fft(opr.wfc, opr.size, true);

        for (size_t j = 0; j < opr.size; ++j) {
            opr.wfc[j] *= opr.pe[j];
        }

        for (size_t j = 0; j < opr.size; ++j) {
            density[j] = pow(cabs(opr.wfc[j]), 2);
        }

        if (par.im_time) {
            double sum = 0;

            for (size_t j = 0; j < opr.size; ++j) {
                sum += density[j];
            }

            sum *= par.dx;

            for (size_t j = 0; j < opr.size; ++j) {
                opr.wfc[j] /= sqrt(sum);
            }
        }

        // Writing data into a file in the format of:
        // index, density, real potential.
        char filename[256];
        sprintf(filename, "output%lu.dat", i);
        FILE *fp = fopen(filename, "w");

        for (int i = 0; i < opr.size; ++i) {
            fprintf(fp, "%d\t%f\t%f\n", i, density[i], creal(opr.v[i]));
        }

        fclose(fp);
    }
}

double calculate_energy(struct params par, struct operators opr) {
    double complex wfc_r[opr.size];
    double complex wfc_k[opr.size];
    double complex wfc_c[opr.size];
    memcpy(wfc_r, opr.wfc, sizeof(wfc_r));

    memcpy(wfc_k, opr.wfc, sizeof(wfc_k));
    fft(wfc_k, opr.size, false);

    for (size_t i = 0; i < opr.size; ++i) {
        wfc_c[i] = conj(wfc_r[i]);
    }

    double complex energy_k[opr.size];
    double complex energy_r[opr.size];

    for (size_t i = 0; i < opr.size; ++i) {
        energy_k[i] = wfc_k[i] * cpow(par.k[i] + 0.0*I, 2);
    }

    fft(energy_k, opr.size, true);

    for (size_t i = 0; i < opr.size; ++i) {
        energy_k[i] *= 0.5 * wfc_c[i];
        energy_r[i] = wfc_c[i] * opr.v[i] * wfc_r[i];
    }

    double energy_final = 0;

    for (size_t i = 0; i < opr.size; ++i) {
        energy_final += creal(energy_k[i] + energy_r[i]);
    }

    return energy_final * par.dx;
}

void free_params(struct params par) {
    free(par.x);
    free(par.k);
}

void free_operators(struct operators opr) {
    free(opr.v);
    free(opr.pe);
    free(opr.ke);
    free(opr.wfc);
}

int main() {
    struct params par;
    struct operators opr;

    init_params(&par, 5.0, 256, 0.05, 100, true);
    init_operators(&opr, par, 0.0, -1.0);

    split_op(par, opr);

    printf("the energy is %f\n", calculate_energy(par, opr));

    free_params(par);
    free_operators(opr);

    return 0;
}

File: contents/split-operator_method/code/go/split_op.go
package main

import (
	"bufio"
	"fmt"
	"github.com/runningwild/go-fftw/fftw"
	"math"
	"math/cmplx"
	"os"
	"strconv"
)

type params struct {
	xmax      float64
	res       int
	dt        float64
	timesteps int
	dx        float64
	x         []float64
	dk        float64
	k         []float64
	im_time   bool
}

type operators struct {
	size int
	v    []complex128
	pe   []complex128
	ke   []complex128
	wfc  []complex128
}

func fft(x []complex128, inv bool) {
	y := fftw.NewArray(len(x))
	for i := 0; i < len(x); i++ {
		y.Set(i, x[i])
	}

	if inv {
		plan := fftw.NewPlan(y, y, fftw.Backward, fftw.Estimate)
		plan.Execute()
		plan.Destroy()
	} else {
		plan := fftw.NewPlan(y, y, fftw.Forward, fftw.Estimate)
		plan.Execute()
		plan.Destroy()
	}

	for i := 0; i < len(x); i++ {
		x[i] = y.At(i) / cmplx.Sqrt(complex(float64(len(x)), 0))
	}

}

func initParams(xmax, dt float64, res, timesteps int, im bool) params {
	var par params
	par.xmax = xmax
	par.res = res
	par.dt = dt
	par.timesteps = timesteps
	par.dx = 2.0 * xmax / float64(res)
	par.x = make([]float64, res)
	par.dk = math.Pi / xmax
	par.k = make([]float64, res)
	par.im_time = im

	for i := 0; i < res; i++ {
		par.x[i] = xmax/float64(res) - xmax + float64(i)*(2.0*xmax/float64(res))
		if i < res/2 {
			par.k[i] = float64(i) * math.Pi / xmax
		} else {
			par.k[i] = float64(i-res) * math.Pi / xmax
		}
	}
	return par
}

func initOperators(par params, voffset, wfcoffset float64) operators {
	var opr operators
	opr.size = par.res
	opr.v = make([]complex128, par.res)
	opr.pe = make([]complex128, par.res)
	opr.ke = make([]complex128, par.res)
	opr.wfc = make([]complex128, par.res)

	for i := 0; i < par.res; i++ {
		opr.v[i] = 0.5 * cmplx.Pow(complex(par.x[i]-voffset, 0), 2)
		opr.wfc[i] = cmplx.Exp(-cmplx.Pow(complex(par.x[i]-wfcoffset, 0), 2) / 2)

		if par.im_time {
			opr.ke[i] = cmplx.Exp(-0.5 * complex(par.dt, 0) * cmplx.Pow(complex(par.k[i], 0), 2))
			opr.pe[i] = cmplx.Exp(-0.5 * complex(par.dt, 0) * opr.v[i])
		} else {
			opr.ke[i] = cmplx.Exp(complex(0, -0.5*par.dt*math.Pow(par.k[i], 2)))
			opr.pe[i] = cmplx.Exp(complex(-imag(opr.v[i]), -0.5*par.dt*real(opr.v[i])))
		}
	}
	return opr
}

func splitOp(par params, opr operators) {
	density := make([]float64, opr.size)

	for i := 0; i < par.timesteps; i++ {
		for j := 0; j < opr.size; j++ {
			opr.wfc[j] *= opr.pe[j]
		}

		fft(opr.wfc, false)

		for j := 0; j < opr.size; j++ {
			opr.wfc[j] *= opr.ke[j]
		}

		fft(opr.wfc, true)

		for j := 0; j < opr.size; j++ {
			opr.wfc[j] *= opr.pe[j]
		}

		for j := 0; j < opr.size; j++ {
			density[j] = math.Pow(cmplx.Abs(opr.wfc[j]), 2)
		}

		if par.im_time {
			sum := 0.0

			for j := 0; j < opr.size; j++ {
				sum += density[j]
			}

			sum *= par.dx

			for j := 0; j < opr.size; j++ {
				opr.wfc[j] /= complex(math.Sqrt(sum), 0)
			}
		}

		filename := "output" + strconv.Itoa(i) + ".dat"
		file, err := os.Create(filename)

		if err != nil {
			fmt.Println("Cant write: ", err)
			continue
		}
		defer file.Close()

		writer := bufio.NewWriter(file)

		fmt.Fprintf(writer, "%d\t%f\t%f\n", i, density[i], real(opr.v[i]))
	}
}

func calculateEnergy(par params, opr operators) float64 {
	wfc_r := make([]complex128, opr.size)
	wfc_k := make([]complex128, opr.size)
	wfc_c := make([]complex128, opr.size)

	copy(wfc_r, opr.wfc)
	copy(wfc_k, opr.wfc)
	fft(wfc_k, false)

	for i := 0; i < opr.size; i++ {
		wfc_c[i] = cmplx.Conj(wfc_r[i])
	}

	energy_r := make([]complex128, opr.size)
	energy_k := make([]complex128, opr.size)

	for i := 0; i < opr.size; i++ {
		energy_k[i] = wfc_k[i] * cmplx.Pow(complex(par.k[i], 0), 2)
	}

	fft(energy_k, true)

	for i := 0; i < opr.size; i++ {
		energy_k[i] *= 0.5 * wfc_c[i]
		energy_r[i] = wfc_c[i] * opr.v[i] * wfc_r[i]
	}

	energy_final := 0.0

	for i := 0; i < opr.size; i++ {
		energy_final += real(energy_k[i] + energy_r[i])
	}

	return energy_final * par.dx
}

func main() {
	par := initParams(5.0, 0.05, 256, 100, true)
	opr := initOperators(par, 0.0, -1.0)

	splitOp(par, opr)

	fmt.Printf("The energy is %f\n", calculateEnergy(par, opr))
}
